### PR TITLE
[WIP] PartDesign: New Text utility

### PR DESCRIPTION
**Pull request by davidosterberg**
Allow for easy and parametric additive and subractive text features in PartDesign.

Work is very much in progress. No need for detail review at this stage. 

Forum discussion:
https://forum.freecadweb.org/viewtopic.php?f=19&t=53833

image
demo1

What works
- Font selection with QFontComboBox
- Generation of text shapes

Todo:

- [x] Fix segfault when recomputing
- [x] Implement the text location by sketch logic
- [ ] Ship a small selection of fonts with FreeCAD, so that models using those fonts will be platform independent, and distinguish these in the UI.
- [ ] implement subractive, additive, and imprint modes by observing the sign of the input 
- [ ] Filter out non TrueType fonts from QFontComboBox (subclass?)
- [ ] Scale text to exact desired height by means of Bounding box command.

### CODE REVIEW

File: src/Mod/PartDesign/App/FeatureText.cpp [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+    Base.setValue(base.x,base.y,base.z);
+    Axis.setValue(dir.x,dir.y,dir.z);
+}
+
+
+TopoDS_Shape Text::textToShape(const std::string& text, const std::string& fontpath, double height, double track = 0.0)
+{
+    Base::Console().Warning("textToShape was called\n");
+
+    Py::Module mod(PyImport_ImportModule("Part"), true);
```
Comment (davidosterberg) [inline src/Mod/PartDesign/App/FeatureText.cpp:290]: This line segfaults when recomputing attached file
[example.zip](https://github.com/FreeCAD/FreeCAD/files/5776761/example.zip)

File: src/Mod/PartDesign/Gui/TaskTextParameters.cpp [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+}
+
+
+// Based on https://stackoverflow.com/a/64729615
+// Ported to c++
+// Modified to ignore Non-TrueType fonts
+void TaskTextParameters::getFontPaths(void)
+{
+    QStringList font_paths = QStandardPaths::standardLocations(QStandardPaths::FontsLocation);
+    QFontDatabase* db = new QFontDatabase();
```
Comment (chennes) [inline src/Mod/PartDesign/Gui/TaskTextParameters.cpp:403]: Probably no reason to dynamically allocate this, is there?

Comment (davidosterberg) [reply src/Mod/PartDesign/Gui/TaskTextParameters.cpp:403]: This was the best way I could figure out to map a font name to the actual file path that I need. Do you mean that I could precompute this somehow?

Anyway, don't spend too much time reviewing now, as the main functionality is not completed. I actually only made the PR now in order to discuss the segmentation fault. Mentioned above.

Comment (chennes) [reply src/Mod/PartDesign/Gui/TaskTextParameters.cpp:403]: I was just giving it a quick look for the obvious problem children (naked pointers being a major culprit in segfaults). What I mean is that there is no reason to create a pointer to the font database at all: just create the object on the stack:
```suggestion
    QFontDatabase db;
```
(and of course change your -> to . in later code...)

Comment (davidosterberg) [reply src/Mod/PartDesign/Gui/TaskTextParameters.cpp:403]: Ah, of course- Done

**Comment by hyarion** [comment]
I like that this makes it easier to create parametric text, but I am wondering if a specific tool for it is necessary. Wouldn't it be better to create the parametric text in 2d and use pad/pocket to make it 3d? (that way it could be possible to use create 3d text with sweep and partial revolve too.

Is there a forum thread where this feature is being discussed?

**Comment by davidosterberg** [comment]
@hyarion there is no forum thread yet (will create one soon). The 2D contour approach exists in Draft -> ShapeString. This is PartDesign and in PD and IMO it is more appropriate to this WB to create a feature directly on the body.

**Comment by hyarion** [comment]
@davidosterberg ok, thanks I'll continue the discussion there instead

**Comment by davidosterberg** [comment]
Closed, in favor of a Sketcher based solution. See forum discussion

**Pull request closed without merging**